### ISSUE TITLE
Unable to use stack transformation from unit test

### ISSUE THREAD

**Issue by clstokes** [pulumi/pulumi#4669]
If I call `pulumi.runtime.registerStackTransformation` from my Pulumi application and run a unit test against my application, the tests fail to run with the error `The root stack resource was referenced before it was initialized.` The same Pulumi application runs successfully without error with `pulumi preview`.

#### Steps to reproduce
1. `git clone https://gist.github.com/20db9640bc9f3c8607f6d67d8390fb45 gh-test && cd gh-test`
1. `npm i`
1. `pulumi stack init gh-test`
1. `npm test`
    - Observe error.
1. `pulumi preview`
    - Observe successful preview.

**Comment by lukehoban** [comment]
This happens because we do not currently create a root `Stack` resource automatically in test mode.  It is likely that we should, and that `setMocks` should initialize a root `Stack` automatically .  Coincidentally, we are looking at making that change for Python for other reasons in https://github.com/pulumi/pulumi/pull/4670, but it is likely we should make that same change in Node.js which we fix this issue.

In the meantime, a workaround is to use `pulumi.runtime.runInPulumiStack` to construct a Stack resource.

```ts
pulumi.runtime.setMocks({
    call: x => {
        return {};
    },
    newResource: x => {
        return {id: "", state: {}};
    }
})
pulumi.runtime.runInPulumiStack(async () => {
    pulumi.runtime.registerStackTransformation(args => {
        console.log(`transofming: ${args.name}`)
        return {
            opts: args.opts,
            props: args.props,
        }
    })
    
    // Create an AWS resource (S3 Bucket)
    const bucket = new aws.s3.Bucket("my-bucket");
    
    // Export the name of the bucket
    return {
        bucketName: bucket.id
    };
})
```

**Comment by amichel** [comment]
This problem scope is bigger than just unit tests. I have common transformations that I want to be able to reuse in all projects as npm package. It also causes problem the other way around, when I register stack transformation, it doesn't apply on resources created by modules from shared npm packages.

**Comment by clstokes** [comment]
Related - https://github.com/pulumi/pulumi/issues/6214

**Comment by arwilczek90** [comment]
I agree with @amichel I get the same traceback when running pulumi up on a project with a stack transformation.
